sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/TracerProvider.java

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.implementation;
+
+import com.azure.core.util.Context;
+import com.azure.core.util.FluxUtil;
+import com.azure.core.util.tracing.Tracer;
+import com.azure.cosmos.CosmosClientException;
+import com.azure.cosmos.models.CosmosAsyncItemResponse;
+import com.azure.cosmos.models.CosmosResponse;
+import com.azure.cosmos.models.ResourceWrapper;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import static com.azure.core.util.tracing.Tracer.AZ_TRACING_NAMESPACE_KEY;
+
+public class TracerProvider {
+    private final List<Tracer> tracers = new ArrayList<>();
+    private final boolean isEnabled;
+    public final static String DB_TYPE_VALUE = "cosmosdb";
+    public final static String DB_TYPE = "db.type";
+    public final static String DB_INSTANCE = "db.instance";
+    public final static String DB_URL = "db.url";
+    public static final String DB_STATEMENT = "db.statement";
+    public static final String ERROR_MSG = "error.msg";
+    public static final String ERROR_TYPE = "error.type";
+    public static final String ERROR_STACK = "error.stack";
+    public static final String COSMOS_CALL_DEPTH = "cosmosCallDepth";
+    public static final String MICROSOFT_DOCOMENTDB = "Microsoft.DocumentDB";
+
+
+    public final static Function<reactor.util.context.Context, reactor.util.context.Context> callDepthAttributeFunc = (
+        reactor.util.context.Context reactorContext) -> {
+        CallDepth callDepth = reactorContext.getOrDefault(COSMOS_CALL_DEPTH, CallDepth.ZERO);
+        assert callDepth != null;
```

Comment:
`callDepth` cannot be null as you are doing `getOrDefault` in the previous line. 

Also, it's better to use `if (callDepth == null) { throw exception; }` instead of using `assert`. If the assertion fails, it throws `AssertionError` extends from `Error` which is considered fatal and the program cannot recover.

Reply:
as discussed offline , it is due to spot bug